Manage client SSL secrets for external DB connectivity (fixes #1088)

mqufflc opened the pull request:
##### SUMMARY

This commit adds a way to provide an SSL  client certificate + private key to authenticate against an external postgresql database.

fixes #1088

##### ISSUE TYPE
 - New or Enhanced Feature


##### ADDITIONAL INFORMATION
The user will have to add the certificate + key in a tls secret and provide it in the awx configuration.

The certificate + key are mounted and copied in order to fix file authorization issues that I have encountered when mounting directly in the pod.

mqufflc commented:
I have been able to deploy an AWX instance using an external postgres database with TLS enforced and `clientcert=verify-full`.

But I still have these logs in my postgres instance :
```
2023-04-28 18:31:07.945 CEST [8607] postgres@awx FATAL:  connection requires a valid client certificate
2023-04-28 18:31:10.634 CEST [8608] postgres@awx FATAL:  connection requires a valid client certificate
2023-04-28 18:31:16.738 CEST [8610] postgres@awx FATAL:  connection requires a valid client certificate
2023-04-28 18:31:25.146 CEST [8573] postgres@awx LOG:  could not receive data from client: Connection reset by peer
2023-04-28 18:31:27.429 CEST [8636] postgres@awx FATAL:  connection requires a valid client certificate
2023-04-28 18:31:33.597 CEST [8642] postgres@awx FATAL:  connection requires a valid client certificate
2023-04-28 18:31:58.350 CEST [8723] postgres@awx FATAL:  connection requires a valid client certificate
2023-04-28 18:32:04.535 CEST [8738] postgres@awx FATAL:  connection requires a valid client certificate
```
I must have forgot a container where the client certificate is needed.

mqufflc commented:
I had forgotten to mount the TLS credentials to the rsyslog containers. Now there is no more errors related to client certificate in the postgres database logs.

If anyone can review my PR and tell me if I need to fix some things, I will be happy to.

jravetch commented:
:+1:  Would love to get this merged and released.  I am also having issues using GCP Cloud SQL with SSL for AWX

jravetch commented:
@fosterseth Any chance you can review this pull request?

fosterseth commented:
@mqufflc I can test this out, can you please rebase and fix the merge conflicts?

mqufflc commented:
I have rebased and fixed the conflicts. 
I haven't got the time to test again external database connectivity with client certificate enforced. I will try again in the following days but I expect it to work the same.

mqufflc commented:
I was able to start an AWX instance and run a job. The database was configured with `clientcert=verify-full` and md5 auth.

TheRealHaoLiu commented:
Hi @mqufflc this is a fantastic contribution and we definitely want to get this into the project. Sorry for the delay, hopefully we will get this this PR soon 

one thing that would help with verification of this PR is if you are able to provide detail instruction (and result) on how you tested your PR.

TheRealHaoLiu commented:
btw the PR is missing descriptor in clusterserviceversion for the operator

here's a example https://github.com/ansible/awx-operator/blob/devel/config/manifests/bases/awx-operator.clusterserviceversion.yaml#L289-L294

Commit:
Manage client SSL secrets for external DB connectivity

mqufflc commented:
Hello @TheRealHaoLiu thank you for your feedback.

I have added the new fields in clusterserviceversion. Let me know if it is good enough.

I have created the following repository containing the kustomize configuration I have used to deploy my AWX instance : [mqufflc/awx-operator-verify-cert](https://github.com/mqufflc/awx-operator-verify-cert)

It also contains some intructions to generate some certificates and how to configure a postgres instance to use them.

I have installed a postgres instance on a host and deployed AWX on a kubernetes cluster. With this configuration I was able to start AWX. The web UI seems to be working fine.

The DB connections are encrypted : 

```
pid  | ssl | version |         cipher         | bits | compression |                         client_dn                          |                  client_serial                  |                         issuer_dn
------+-----+---------+------------------------+------+-------------+------------------------------------------------------------+-------------------------------------------------+-----------------------------------------------------------
 2297 | t   | TLSv1.3 | TLS_AES_256_GCM_SHA384 |  256 | f           | /C=AU/ST=Some-State/O=Internet Widgits Pty Ltd/CN=postgres | 99070950598692336244154921773583965827850112992 | /C=AU/ST=Some-State/O=Internet Widgits Pty Ltd/CN=psql-ca
 2298 | t   | TLSv1.3 | TLS_AES_256_GCM_SHA384 |  256 | f           | /C=AU/ST=Some-State/O=Internet Widgits Pty Ltd/CN=postgres | 99070950598692336244154921773583965827850112992 | /C=AU/ST=Some-State/O=Internet Widgits Pty Ltd/CN=psql-ca
 2299 | t   | TLSv1.3 | TLS_AES_256_GCM_SHA384 |  256 | f           | /C=AU/ST=Some-State/O=Internet Widgits Pty Ltd/CN=postgres | 99070950598692336244154921773583965827850112992 | /C=AU/ST=Some-State/O=Internet Widgits Pty Ltd/CN=psql-ca
 2324 | t   | TLSv1.3 | TLS_AES_256_GCM_SHA384 |  256 | f           | /C=AU/ST=Some-State/O=Internet Widgits Pty Ltd/CN=postgres | 99070950598692336244154921773583965827850112992 | /C=AU/ST=Some-State/O=Internet Widgits Pty Ltd/CN=psql-ca
 2334 | t   | TLSv1.3 | TLS_AES_256_GCM_SHA384 |  256 | f           | /C=AU/ST=Some-State/O=Internet Widgits Pty Ltd/CN=postgres | 99070950598692336244154921773583965827850112992 | /C=AU/ST=Some-State/O=Internet Widgits Pty Ltd/CN=psql-ca
 2330 | t   | TLSv1.3 | TLS_AES_256_GCM_SHA384 |  256 | f           | /C=AU/ST=Some-State/O=Internet Widgits Pty Ltd/CN=postgres | 99070950598692336244154921773583965827850112992 | /C=AU/ST=Some-State/O=Internet Widgits Pty Ltd/CN=psql-ca
 2807 | t   | TLSv1.3 | TLS_AES_256_GCM_SHA384 |  256 | f           | /C=AU/ST=Some-State/O=Internet Widgits Pty Ltd/CN=postgres | 99070950598692336244154921773583965827850112992 | /C=AU/ST=Some-State/O=Internet Widgits Pty Ltd/CN=psql-ca
 2791 | t   | TLSv1.3 | TLS_AES_256_GCM_SHA384 |  256 | f           | /C=AU/ST=Some-State/O=Internet Widgits Pty Ltd/CN=postgres | 99070950598692336244154921773583965827850112992 | /C=AU/ST=Some-State/O=Internet Widgits Pty Ltd/CN=psql-ca
 2805 | t   | TLSv1.3 | TLS_AES_256_GCM_SHA384 |  256 | f           | /C=AU/ST=Some-State/O=Internet Widgits Pty Ltd/CN=postgres | 99070950598692336244154921773583965827850112992 | /C=AU/ST=Some-State/O=Internet Widgits Pty Ltd/CN=psql-ca
 2806 | t   | TLSv1.3 | TLS_AES_256_GCM_SHA384 |  256 | f           | /C=AU/ST=Some-State/O=Internet Widgits Pty Ltd/CN=postgres | 99070950598692336244154921773583965827850112992 | /C=AU/ST=Some-State/O=Internet Widgits Pty Ltd/CN=psql-ca
 2790 | t   | TLSv1.3 | TLS_AES_256_GCM_SHA384 |  256 | f           | /C=AU/ST=Some-State/O=Internet Widgits Pty Ltd/CN=postgres | 99070950598692336244154921773583965827850112992 | /C=AU/ST=Some-State/O=Internet Widgits Pty Ltd/CN=psql-ca
 2808 | t   | TLSv1.3 | TLS_AES_256_GCM_SHA384 |  256 | f           | /C=AU/ST=Some-State/O=Internet Widgits Pty Ltd/CN=postgres | 99070950598692336244154921773583965827850112992 | /C=AU/ST=Some-State/O=Internet Widgits Pty Ltd/CN=psql-ca
 2833 | f   |         |                        |      |             |                                                            |                                                 |
(13 rows)
```

I have also been able to launch some playbooks from the web UI, create a user and log in as this user.

I have changed my AWX configuration (deleting the `postgres_client_ssl_secret` field in awx.yaml specs) to check if my postgres instance was not allowing connections without certificate.

As a result AWX stopped working and I got these logs on my postgres instance : 

```
2023-08-09 22:27:52.487 CEST [2905] postgres@awx FATAL:  connection requires a valid client certificate
2023-08-09 22:27:53.601 CEST [2907] postgres@awx FATAL:  connection requires a valid client certificate
2023-08-09 22:27:53.875 CEST [2908] postgres@awx FATAL:  connection requires a valid client certificate
2023-08-09 22:27:54.204 CEST [2909] postgres@awx FATAL:  connection requires a valid client certificate
2023-08-09 22:27:54.647 CEST [2910] postgres@awx FATAL:  connection requires a valid client certificate
2023-08-09 22:27:55.646 CEST [2911] postgres@awx FATAL:  connection requires a valid client certificate
2023-08-09 22:27:56.157 CEST [2912] postgres@awx FATAL:  connection requires a valid client certificate
2023-08-09 22:27:56.645 CEST [2913] postgres@awx FATAL:  connection requires a valid client certificate
2023-08-09 22:27:57.347 CEST [2914] postgres@awx FATAL:  connection requires a valid client certificate
```

Bringing back the client certificate secret in AWX specs resolved the issue, as expected.

Let me know if you want other results and which one if needed.

farr-slmn commented:
@TheRealHaoLiu Hello, just pinging as this solution from @mqufflc is crucial for us users🙏. Thank you for your time!

Cecil160 commented:
We, too, could benefit substantially from this PR as we need to deploy AWX in GKE with SSL-enforced CloudSQL.

DE110283 commented:
+1
